File: android/src/main/java/com/theoplayer/broadcast/DefaultEventDispatcher.kt
package com.theoplayer.broadcast

import com.theoplayer.android.api.event.EventDispatcher
import com.theoplayer.android.api.event.EventListener
import com.theoplayer.android.api.event.EventType
import com.theoplayer.android.api.event.Event

open class DefaultEventDispatcher: EventDispatcher<Event<*>> {
    private val _listeners = mutableMapOf<EventType<*>, MutableList<EventListener<*>>>()

    override fun <T : Event<*>> addEventListener(eventType: EventType<T>, listener: EventListener<in T>) {
        if (!_listeners.contains(eventType)) {
            _listeners[eventType] = mutableListOf()
        }
        _listeners[eventType]?.add(listener)
    }

    override fun <T : Event<*>> removeEventListener(eventType: EventType<T>, listener: EventListener<in T>) {
       _listeners[eventType]?.remove(listener)
    }

    fun dispatchEvent(event: Event<*>) {
        _listeners[event.type]?.forEach { listener ->
          @Suppress("UNCHECKED_CAST")
          (listener as EventListener<Event<*>>).handleEvent(event)
        }
    }
}

File: src/internal/adapter/event/DefaultEventDispatcher.ts
import type { EventDispatcher, EventMap, StringKeyOf } from 'react-native-theoplayer';
import type { EventListener } from 'react-native-theoplayer';
import { arrayRemoveElement } from '../../utils/arrayUtil';

export class DefaultEventDispatcher<TMap extends EventMap<StringKeyOf<TMap>>> implements EventDispatcher<TMap> {
  readonly _eventListeners: Map<StringKeyOf<TMap>, EventListener<TMap[StringKeyOf<TMap>]>[]> = new Map();

  addEventListener<K extends StringKeyOf<TMap>>(type: K | readonly K[], listener: EventListener<TMap[K]>): void {
    if (typeof listener !== 'function') {
      return;
    } else if (Array.isArray(type)) {
      type.forEach((t) => this.addSingleEventListener(t, listener));
    } else {
      this.addSingleEventListener(type as K, listener);
    }
  }

  private addSingleEventListener<K extends StringKeyOf<TMap>>(type: K, listener: EventListener<TMap[K]>): void {
    if (!this._eventListeners.has(type)) {
      this._eventListeners.set(type, [listener as EventListener<TMap[StringKeyOf<TMap>]>]);
    } else {
      this._eventListeners.get(type)?.push(listener as EventListener<TMap[StringKeyOf<TMap>]>);
    }
  }

  clearEventListeners(): void {
    this._eventListeners.clear();
  }

  dispatchEvent = <K extends StringKeyOf<TMap>>(event: TMap[K]): void => {
    const listeners = (this._eventListeners.get(event.type) ?? []).slice();
    for (const listener of listeners) {
      listener.call(this, event);
    }
  };

  removeEventListener<K extends StringKeyOf<TMap>>(type: K | readonly K[], listener: EventListener<TMap[K]>): void {
    if (typeof listener !== 'function') {
      return;
    } else if (Array.isArray(type)) {
      type.forEach((t) => this.removeSingleEventListener(t, listener));
    } else {
      this.removeSingleEventListener(type as K, listener);
    }
  }

  private removeSingleEventListener<K extends StringKeyOf<TMap>>(type: K, listener: EventListener<TMap[K]>): void {
    const listeners = this._eventListeners.get(type);
    if (listeners) {
      arrayRemoveElement(listeners, listener);
    }
  }
}
